### PR TITLE
Improved UX for sweeping accounts

### PR DESCRIPTION
**Pull request by EtDu**
This fix aims to solve the problem of accounts leaving dust when swept after a gas price change. 

The idea is to make 'max mode' a toggle feature which disables the send amount input box and allows the user to change the gas, which would then update the send amount accordingly.

### CODE REVIEW

**Comment by ContemporaryAnarchy** [comment]
FeatureDemo

**Comment by ContemporaryAnarchy** [comment]
FeatureDemo

**Comment by ContemporaryAnarchy** [comment]
Still need to adjust for the advanced options

**Comment by ContemporaryAnarchy** [comment]
updatedDemo
**UPDATED

**Comment by bdresser** [comment]
thanks @ContemporaryAnarchy, this is a great fix. We're going to have our designer take a look at the `Max` button to keep it in line with our style guide, will get back to you here shortly!

**Comment by ContemporaryAnarchy** [comment]
> thanks @ContemporaryAnarchy, this is a great fix. We're going to have our designer take a look at the `Max` button to keep it in line with our style guide, will get back to you here shortly!

My pleasure. Looking forward to feedback for the implementation

**Comment by ContemporaryAnarchy** [comment]
Checked over the integration tests with @kumavis , couldn't find the solution, perhaps someone can shed some light on them 👍

**Comment by bdresser** [comment]
hey @ContemporaryAnarchy! our designer put together this mockup to make sure it stays in sync with the rest of the extension design

<a href="https://cl.ly/8dff77178a86" target="_blank"></a>

Link to [Figma](https://www.figma.com/proto/Ko47bJeCdyCKsePa2ObWGjln/Send-Screen?node-id=1%3A367&viewport=461%2C169%2C0.651491&scaling=min-zoom)

are you willing to implement this visual treatment, or should we have someone on our side take it over the finish line!

thanks again for your work

**Comment by bdresser** [comment]
@tmashuang any clue on the tests?

**Comment by ContemporaryAnarchy** [comment]
Thanks @bdresser, I'll implement it. There isn't enough space for the insufficient funds message below the box, how should I handle that? I'm thinking the 'from' box could stand to be more narrow
Screen Shot 2019-05-09 at 10 43 52 AM

**Comment by ContemporaryAnarchy** [comment]
This is the result, let me know what you think
metamask-new-max-button

**Comment by bdresser** [comment]
@cjeria wdyt ? ^

**Comment by cjeria** [comment]
Looks good! Two things though:

- Let's omit the "max set" label under the input field. I think it's obvious enough with the blue button/label and I'm not liking how the last field jumps when toggling the max option.
- When the max is set, give the input field a bg-color: #F2F3F4; to emphasize that it's not editable.

@bdresser

**Comment by ContemporaryAnarchy** [comment]
Fixed existing integration tests and added some more, feel free to merge unless any further inspection is necessary :)

File: ui/app/pages/send/send-content/send-row-wrapper/send-row-wrapper.component.js [hunk 1]
```diff
@@ -45,4 +77,18 @@ export default class SendRowWrapper extends Component {
...
+  render () {
+    const {
+      errorType = '',
+    } = this.props
+
+    return (
+      errorType === 'amount'
+      ?
+      this.renderAmountFormRow()
+      :
```
Comment (whymarrh) [inline ui/app/pages/send/send-content/send-row-wrapper/send-row-wrapper.component.js]: Can we put this onto one line?

```js
errorType === 'amount' ? this.renderAmountFormRow() : this.renderFormRow()
```

File: ui/app/pages/send/send-content/send-row-wrapper/send-row-wrapper.component.js [hunk 2]
```diff
@@ -34,7 +34,39 @@ export default class SendRowWrapper extends Component {
...
+      children,
+      errorType = '',
+      label,
+      showError = false,
+      showWarning = false,
+      warningType = '',
+    } = this.props
+
+    const formField = Array.isArray(children) ? children[1] || children[0] : children
+    const customLabelContent = children.length > 1 ? children[0] : null
```
Comment (whymarrh) [inline ui/app/pages/send/send-content/send-row-wrapper/send-row-wrapper.component.js]: If we're uncertain that children is an array, we should include that in the condition here:

```suggestion
    const customLabelContent = (Array.isArray(children) && children.length) > 1 ? children[0] : null
```

Comment (whymarrh) [reply ui/app/pages/send/send-content/send-row-wrapper/send-row-wrapper.component.js]: In which case we might as well introduce an if statement to handle the combined cases

File: ui/app/pages/send/send-content/send-gas-row/send-gas-row.component.js [hunk 3]
```diff
@@ -89,7 +121,17 @@ export default class SendGasRow extends Component {
       convertedCurrency={convertedCurrency}
       gasLoadingError={gasLoadingError}
       gasTotal={gasTotal}
-      onReset={resetGasButtons}
+      onReset={() => {
+        return new Promise((resolve, reject) => {
```
Comment (whymarrh) [inline ui/app/pages/send/send-content/send-gas-row/send-gas-row.component.js]: The function passed to the `Promise` constructor is executed synchronously and we're not doing anything async in the function so we can leave it out. We should be able to do something like:

```js
onReset={() => {
  resetGasButtons()
  if (maxModeOn) {
    this.setMaxAmount()
  }
}}
```

File: ui/app/pages/send/send-content/send-gas-row/send-gas-row.component.js [hunk 4]
```diff
@@ -79,7 +102,16 @@ export default class SendGasRow extends Component {
                 name: 'Changed Gas Button',
               },
             })
-            gasPriceButtonGroupProps.handleGasPriceSelection(...args)
+            return new Promise((resolve, reject) => {
```
Comment (whymarrh) [inline ui/app/pages/send/send-content/send-gas-row/send-gas-row.component.js]: The function passed to the `Promise` constructor is executed synchronously and we're not doing anything async in the function so we can leave it out. We should be able to do something like:

```js
gasPriceButtonGroupProps.handleGasPriceSelection(...args)
if (maxModeOn) {
  this.setMaxAmount()
}
```

File: ui/app/css/itcss/components/send.scss [hunk 5]
```diff
@@ -781,12 +791,47 @@
...
+      bottom: 0;
+      display: flex;
+      align-items: center;
+      justify-content: center;
+      color: #2f9ae0;
+
+      &__disabled {
+      color: #B0D7F2;
+      cursor: auto;
+    }
```
Comment (whymarrh) [inline ui/app/css/itcss/components/send.scss]: Nit: this formatting is a bit off

File: ui/app/components/app/gas-customization/gas-modal-page-container/gas-modal-page-container.container.js [hunk 6]
```diff
@@ -196,18 +232,26 @@ const mergeProps = (stateProps, dispatchProps, ownProps) => {
...
-      } else if (isSpeedUp) {
-        dispatchCreateSpeedUpTransaction(txId, gasPrice)
-        dispatchHideSidebar()
-        dispatchCancelAndClose()
-      } else {
-        dispatchSetGasData(gasLimit, gasPrice)
-        dispatchHideGasButtonGroup()
-        dispatchCancelAndClose()
-      }
+        if (isConfirm) {
```
Comment (whymarrh) [inline ui/app/components/app/gas-customization/gas-modal-page-container/gas-modal-page-container.container.js]: Nit: this whole fn has been over-indented. Can we de-indent this one level to correct the diff?

File: ui/app/components/app/gas-customization/gas-modal-page-container/gas-modal-page-container.container.js [hunk 7]
```diff
@@ -98,9 +107,25 @@ const mapStateToProps = (state, ownProps) => {
   const isMainnet = getIsMainnet(state)
   const showFiat = Boolean(isMainnet || showFiatInTestnets)
 
-  const insufficientBalance = !isBalanceSufficient({
+  const newTotalEth = maxModeOn
+    ?
```
Comment (whymarrh) [inline ui/app/components/app/gas-customization/gas-modal-page-container/gas-modal-page-container.container.js]: Nit: can we put this ternary expression on one line

File: ui/app/components/app/gas-customization/gas-modal-page-container/gas-modal-page-container.container.js [hunk 8]
```diff
@@ -98,9 +107,25 @@ const mapStateToProps = (state, ownProps) => {
...
+    addHexWEIsToRenderableEth(balance, '0x0')
+    :
+    addHexWEIsToRenderableEth(value, customGasTotal)
+
+  const sendAmount = maxModeOn
+    ?
```
Comment (whymarrh) [inline ui/app/components/app/gas-customization/gas-modal-page-container/gas-modal-page-container.container.js]: Nit: can we put this ternary expression on one line

File: ui/app/components/app/gas-customization/gas-modal-page-container/gas-modal-page-container.container.js [hunk 9]
```diff
@@ -98,9 +107,25 @@ const mapStateToProps = (state, ownProps) => {
...
+    subtractHexWEIsFromRenderableEth(balance, customGasTotal)
+    :
+    addHexWEIsToRenderableEth(value, '0x0')
+
+  const insufficientBalance = maxModeOn
+  ?
```
Comment (whymarrh) [inline ui/app/components/app/gas-customization/gas-modal-page-container/gas-modal-page-container.container.js]: Nit: can we put this ternary expression on one line

**Review by whymarrh — CHANGES_REQUESTED** [review]

File: ui/app/pages/send/send-content/send-row-wrapper/send-row-wrapper.component.js (hunk 2 above)
Comment (ContemporaryAnarchy) [reply ui/app/pages/send/send-content/send-row-wrapper/send-row-wrapper.component.js]: This one I didn't change, it must have just moved it, but thank you for the comment!

File: ui/app/pages/send/send-content/send-gas-row/send-gas-row.component.js (hunk 3 above)
Comment (ContemporaryAnarchy) [reply ui/app/pages/send/send-content/send-gas-row/send-gas-row.component.js]: Awesome, there were other parts in which I used a promise that were not necessary, thank you

File: ui/app/pages/send/send-content/send-gas-row/send-gas-row.component.js (hunk 4 above)
Comment (ContemporaryAnarchy) [reply ui/app/pages/send/send-content/send-gas-row/send-gas-row.component.js]: For this one, I got rid of the promise and replaced with simple async/await, else it tries to set the max amount prematurely and causes an insufficient funds error.

**Comment by ContemporaryAnarchy** [comment]
@whymarrh Changes have been implemented, thank you for your feedback 👍

**Review by whymarrh — APPROVED** [review]
This LGTM, the tests have been updated and the implementation looks fine.

Thanks @ContemporaryAnarchy!

**Pull request merged**